core/src/main/java/io/questdb/griffin/engine/orderby/LimitedSizeLongTreeChain.java

```diff
@@ -168,67 +166,71 @@ public void print(Utf16Sink sink, ValuePrinter printer) {
         }
     }
 
+    // important invariant:
+    // when (maxValues == currentValues) then upon returning from this method the comparator left side must be set
+    // to the moreStableRecord with the max/min rowId
     public void put(
-            Record leftRecord,
+            Record currentRecord,
             RecordCursor sourceCursor,
-            Record rightRecord,
+            Record moreStableRecord,
             RecordComparator comparator
     ) {
         if (maxValues == 0) {
             return;
         }
 
-        comparator.setLeft(leftRecord);
-
         // if maxValues < 0 then there's no limit (unless there's more than 2^64 records, which is unlikely)
         if (maxValues == currentValues) {
-            if (rightRecordRowId != minMaxRowId) {
-                sourceCursor.recordAt(rightRecord, minMaxRowId);
-                rightRecordRowId = minMaxRowId;
-            }
-            int cmp = comparator.compare(rightRecord);
+            int cmp = comparator.compare(currentRecord);
 
-            if (isFirstN && cmp >= 0) { // bigger than max for firstN/bottomN
+            if (isFirstN && cmp <= 0) { // bigger than max for firstN/bottomN
                 return;
-            } else if (!isFirstN && cmp <= 0) { // smaller than min for lastN/topN
+            } else if (!isFirstN && cmp >= 0) { // smaller than min for lastN/topN
                 return;
             } else { // record has to be inserted, so we've to remove current minMax
                 removeAndCache(minMaxNode);
             }
         }
 
         if (root == EMPTY) {
-            putParent(leftRecord.getRowId());
+            long currentRecordRowId = currentRecord.getRowId();
+            putParent(currentRecordRowId);
             minMaxNode = root;
-            minMaxRowId = leftRecord.getRowId();
+            minMaxRowId = currentRecordRowId;
             currentValues++;
+            prepareComparatorIfNeeded(sourceCursor, moreStableRecord, comparator);
             return;
         }
 
+        // ok, we need to insert new record into already existing tree
+        // let's optimize for tree-traversal
+        comparator.setLeft(currentRecord);
```

Comment:
It's not that simple. When performing the initial check in the `if (maxValues == currentValues)` branch, minMax is set as the left side of a comparator. We assume we will exit the `put()` method from this branch. If this assumption holds, we can check many records without calling `comparator.setLeft()`.

However, if we do not exit the method from the `if (maxValues == currentValues)` branch, we then start traversing the tree with stored rowIds. In this case, it's better to have the 'current record' as the left side, since this record does not change while traversing the tree.

Does it make sense?

Reply:
Yup, it makes sense. My point is that `moreStableRecord` may be not so stable, so the name is a bit confusing. But if you prefer to keep the naming as is, I'm more than fine with that.

Reply:
how about `ownedRecord`? since this record is owned by the tree-chain while the other is managed externally.

Reply:
Yeah, that's a better name.